repl/src/config_params.rs

```diff
@@ -0,0 +1,44 @@
+
+use std::path::{Path, PathBuf};
+
+#[derive(Default, Debug)]
+pub struct ReplParams {
+    config_dir: PathBuf,
+    include_paths: Vec<PathBuf>,
+    pub history_file: Option<PathBuf>,
+}
+
+impl ReplParams {
+    pub fn from_config_dir(path: &Path) -> Self {
+
+        //Create the modules dir inside the config dir, if it doesn't already exist
+        let modules_dir = path.join("modules");
+        std::fs::create_dir_all(&modules_dir).unwrap();
+
+        Self {
+            config_dir: path.into(),
+            include_paths: vec![],
+            history_file: Some(path.join("history.txt")),
+        }
+    }
+    pub fn push_include_paths(&mut self, paths: Vec<PathBuf>) {
+        self.include_paths.extend(paths);
+    }
+
+    /// Returns the search paths, in order to search
+    ///
+    /// The current working directory is always returned first, and the modules directory
+    /// is always last in the search path.  In the middle are the user-supplied search paths
+    pub fn modules_search_paths<'a>(&'a self) -> impl Iterator<Item=PathBuf> + 'a {
+        let cwd = match std::env::current_dir() {
```

Comment:
Loading cwd dynamically means modules paths are changed when current directory is changed. May be we can just form the list of paths on the start and allow user modify them dynamically. We could also allow removing `modules` directory from paths.

Reply:
Interesting.  I thought that was a feature not a bug.  Although I wasn't aware there were `MeTTa` grounded atoms that let the user move around the file system - but I could see this being something which will be written in the future.

Are saying the cwd when the repl is invoked should be the first path searched for imports?  Should the current cwd be searched at all, in your opinion?

Reply:
> We could also allow removing `modules` directory from paths

What would the reason be to not include the `modules` directory?  My thinking is that this is the equivalent to Python's `site-packages` directory.  Basically a place where MeTTa extensions could be installed by the user.

Reply:
> Are saying the cwd when the repl is invoked should be the first path searched for imports? Should the current cwd be searched at all, in your opinion?

It is an interesting question, I find out recently that Python has more complex rules than I expect for this: https://docs.python.org/3/library/sys.html#sys.path For example it doesn't always include cwd into `sys.path` due to safety reasons I believe. We could kind of reuse Python rules:
- include scripts directory when script is called;
- include current directory in interactive mode.

> What would the reason be to not include the modules directory?

I suggest allowing removing this directory but we still can add it on the start. If user can change some environment then we should allow him change every aspect of it. Some users may want to cleanup or rebuild paths completely why should we add some special code to prevent them from doing it.

Reply:
Include scripts directory when script is called makes sense as when user creates `some.metta` file, local `some.py` module is put into the same directory. After that user expects that `metta some.metta` runs successfully and imports `some.py` because these two scripts in the same directory. It doesn't make a big difference for development but it is different when script is used as an application from different parts of the filesystem. 

Including `cwd` exposes a vulnerability when attacker who knows dependencies of the popular script and has write permissions to the directory from which user works (`/tmp` for instance) can add intercepting module into filesystem thus effectively running his own code with user's privileges and user's input while user thinks that only program he started is running. On the other hand when user runs interpreter in interactive mode he is likely developer and understands consequences of his actions.

Reply:
Ok.  Just so I understand, Sensible search order would be:

1.) parent directory of script being run, or starting working dir for interactive mode
2.) any explicitly supplied include paths passed as cmd-line args
3.) any config paths set by the user in the config.metta (currently this isn't implemented but hopefully very soon)
4.) the "modules" dir

Did I understand correctly?  Or did I miss something?

Reply:
Yes, I am not sure if we need to distinguish 3 and 4, but otherwise it is what I am proposing.

Reply:
Behavior updated in 0254521 commit.